### PR TITLE
Rec/quoted quote

### PR DESCRIPTION
**Pull request by recri**
I think this solves https://github.com/JuliaEditorSupport/julia-emacs/issues/142 by simply rewriting `julia-char-regex`.  The regex didn't actually recognize `'\'`' as a character constant as it was written.  It could still be wrong.   Added a test for my particular case.

Sorry, I should have rebased or something to squash  the intermediate commits.

### CODE REVIEW

File: julia-mode.el
Comment (non-Jedi) [inline julia-mode.el:115]: Does this match [e.g. `'\n'`, `\0`, `\x7f`, `\177`](https://docs.julialang.org/en/v1/manual/strings/#man-characters)? It doesn't seem like it would. Probably worth adding a test for those.

File: julia-mode-tests.el
Comment (non-Jedi) [inline julia-mode-tests.el:485]: Please add `julia--should-font-lock` tests for examples from #142.

**Review by non-Jedi — CHANGES_REQUESTED** [review]
If the tests I suggested are added and pass, this LGTM.

File: julia-mode.el
Comment (recri) [reply julia-mode.el:115]: They all fit into the "(1 to 9 (not (backslash or singlequote))) after backslash" clause.  I could be more prescriptive, if you like.

File: julia-mode-tests.el
Comment (non-Jedi) [inline julia-mode-tests.el:637]: Need to test that the string-face ends at the end of the char literal.

File: julia-mode.el
Comment (non-Jedi) [reply julia-mode.el:115]: That makes sense. This is fine. If you wanted to, you could add a couple comments with more explanation of the different forms of char literals and how they're matched, but I don't think that should block merging this.

Comment (tpapp) [reply julia-mode.el:115]: Yes, explanations would be great if @recri has the time. Also, indentation of `(seq ...` looks a bit off.

File: julia-mode.el
Comment (tpapp) [inline julia-mode.el:121]:
```suggestion
        (or                             ; two alternatives
         (not (any "\\'"))              ; one character, not single quote or backslash
         (seq "\\"                      ; sequence of a backslash followed by ...
              (or                       ; four alternatives
               (any "\\'\"?abfnrtv")    ; single character escape
               (repeat 1 3 (any "0-7")) ; octal escape
               (seq 'x' (repeat 1 8 hex))     ; hex escape
               (seq 'u' (repeat 1 4 hex))     ; unicode escape
               (seq 'U' (repeat 1 8 hex)))))  ; extended unicode escaple
```

**Comment by recri** [comment]
Sorry, I seem to have crossed the git gods somehow.

**Comment by tpapp** [comment]
Thanks for all the work! LGTM, I will merge when @non-Jedi approves.

File: julia-mode.el
Comment (non-Jedi) [inline julia-mode.el:114]: `'''` is a valid Char constant for `'\''`. This doesn't match that does it?

Comment (recri) [reply julia-mode.el]: Amazing, that must have been the only quoted single quote that the original code recognized, and the only one that's not documented in the julia manual.

This last pull also tightens up the \x syntax, julia only permits two hex digits, C permits as many as you write though not guaranteeing what will happen to the excess..

**Review by non-Jedi — APPROVED** [review]
Thanks! LGTM.

**Pull request merged**
